File: 3622-check-divisibility-by-digit-sum-and-product/3622-check-divisibility-by-digit-sum-and-product.go
func checkDivisibility(n int) bool {
    sum, prod := findSumProduct(n)

    return n % (sum + prod) == 0
}

func findSumProduct(n int) (int, int) {
    sum, prod := 0, 1
    for n > 0 {
        d := n % 10
        sum += d
        prod *= d
        n /= 10
    }

    return sum, prod
}

// math
// time: O(k)
// space: O(1)

File: 3622-check-divisibility-by-digit-sum-and-product/3622-check-divisibility-by-digit-sum-and-product.py
class Solution:
    def checkDivisibility(self, n: int) -> bool:
        sum, prod = findSumProduct(n)

        return n % (sum + prod) == 0

def findSumProduct(n: int) -> tuple[int, int]:
    sum, prod = 0, 1
    while n > 0:
        d = n % 10
        sum += d
        prod *= d
        n //= 10

    return sum, prod

# math
# time: O(log(n))
# space: O(1)